From 20aa1be7d78ef0d3bd0b42b1b596b2815b1560ac Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sun, 24 Jan 2016 20:54:41 +0000
Subject: [PATCH] [CONFIGURATION-615] Changed generic types in getList()
 method.

The default value passed to getList() is now of type List<?> rather than
List<Object>. This change of the signature has been part of the 1.10
codebase, but it had not been merged to trunk. The corresponding issue
is CONFIGURATION-558.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/configuration/trunk@1726563 13f79535-47bb-0310-9956-ffa450edef68
---
 .../commons/configuration2/AbstractConfiguration.java      | 7 +++++--
 .../commons/configuration2/CompositeConfiguration.java     | 7 +++++--
 .../configuration2/DynamicCombinedConfiguration.java       | 2 +-
 .../commons/configuration2/ImmutableConfiguration.java     | 2 +-
 .../configuration2/PatternSubtreeConfigurationWrapper.java | 2 +-
 5 files changed, 13 insertions(+), 7 deletions(-)

diff --git a/src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java b/src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
index 7642134ea0..f41d632f45 100644
--- a/src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/AbstractConfiguration.java
@@ -1425,7 +1425,7 @@ public List<Object> getList(String key)
     }
 
     @Override
-    public List<Object> getList(String key, List<Object> defaultValue)
+    public List<Object> getList(String key, List<?> defaultValue)
     {
         Object value = getProperty(key);
         List<Object> list;
@@ -1448,7 +1448,10 @@ else if (value instanceof List)
         }
         else if (value == null)
         {
-            list = defaultValue;
+            // This is okay because we just return this list to the caller
+            @SuppressWarnings("unchecked")
+            List<Object> resultList = (List<Object>) defaultValue;
+            list = resultList;
         }
         else if (value.getClass().isArray())
         {
diff --git a/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java b/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java
index e9946de05b..f23eb9ba7d 100644
--- a/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/CompositeConfiguration.java
@@ -380,7 +380,7 @@ protected boolean containsKeyInternal(String key)
     }
 
     @Override
-    public List<Object> getList(String key, List<Object> defaultValue)
+    public List<Object> getList(String key, List<?> defaultValue)
     {
         List<Object> list = new ArrayList<Object>();
 
@@ -400,7 +400,10 @@ public List<Object> getList(String key, List<Object> defaultValue)
 
         if (list.isEmpty())
         {
-            return defaultValue;
+            // This is okay because we just return this list to the caller
+            @SuppressWarnings("unchecked")
+            List<Object> resultList = (List<Object>) defaultValue;
+            return resultList;
         }
 
         ListIterator<Object> lit = list.listIterator();
diff --git a/src/main/java/org/apache/commons/configuration2/DynamicCombinedConfiguration.java b/src/main/java/org/apache/commons/configuration2/DynamicCombinedConfiguration.java
index fe0239c7c2..a4493c01aa 100644
--- a/src/main/java/org/apache/commons/configuration2/DynamicCombinedConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/DynamicCombinedConfiguration.java
@@ -522,7 +522,7 @@ protected Iterator<String> getKeysInternal(String prefix)
     }
 
     @Override
-    public List<Object> getList(String key, List<Object> defaultValue)
+    public List<Object> getList(String key, List<?> defaultValue)
     {
         return this.getCurrentConfig().getList(key, defaultValue);
     }
diff --git a/src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java b/src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
index c552442b1b..83fe8b8f67 100644
--- a/src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
+++ b/src/main/java/org/apache/commons/configuration2/ImmutableConfiguration.java
@@ -533,7 +533,7 @@ public interface ImmutableConfiguration
      *         object that is not a List.
      * @see #getList(Class, String, List)
      */
-    List<Object> getList(String key, List<Object> defaultValue);
+    List<Object> getList(String key, List<?> defaultValue);
 
     /**
      * Get an object of the specified type associated with the given
diff --git a/src/main/java/org/apache/commons/configuration2/PatternSubtreeConfigurationWrapper.java b/src/main/java/org/apache/commons/configuration2/PatternSubtreeConfigurationWrapper.java
index cb571fd1eb..39f8824662 100644
--- a/src/main/java/org/apache/commons/configuration2/PatternSubtreeConfigurationWrapper.java
+++ b/src/main/java/org/apache/commons/configuration2/PatternSubtreeConfigurationWrapper.java
@@ -225,7 +225,7 @@ protected Iterator<String> getKeysInternal(String prefix)
     }
 
     @Override
-    public List<Object> getList(String key, List<Object> defaultValue)
+    public List<Object> getList(String key, List<?> defaultValue)
     {
         return config.getList(makePath(key), defaultValue);
     }
